### PR TITLE
Extract the new base class and adapt all adapters

### PR DESCRIPTION
**Pull request by PandaJunk**
Extract the new base class and adapt all adapters

### CODE REVIEW

File: ts_benchmark/baselines/utils.py [hunk 1]
```diff
@@ -12,6 +13,60 @@
...
+            self.counter += 1
+            print(f"EarlyStopping counter: {self.counter} out of {self.patience}")
+            if self.counter >= self.patience:
+                self.early_stop = True
+        else:
+            self.best_score = score
+            self.save_checkpoint(val_loss, model)
+            self.counter = 0
+
+    def save_checkpoint(self, val_loss, model):
```
Comment (luckiezhou) [inline ts_benchmark/baselines/utils.py:61]: In most deep learning frameworks, save_checkpoint is defined in the trainer class (e.g. look at lightning), if it is placed in EarlyStopping, it may be difficult for users to save the model manually. I recommend to put it in the model class for now, and change if we have a well-designed architecture in the future.

File: ts_benchmark/models/advanced_model_base.py [hunk 2]
```diff
@@ -0,0 +1,724 @@
...
+
+        for key, value in kwargs.items():
+            setattr(self, key, value)
+
+    @property
+    def pred_len(self):
+        return self.horizon
+
+
+class Advanced_Model_Base(ModelBase):
```
Comment (luckiezhou) [inline ts_benchmark/models/advanced_model_base.py:53]: Please follow the naming convention in the benchmark, and rename it to 'AdvancedModelBase'
Moreover, the name should reflect what types of algorithms does this class support, for example, should it be a 'DeepModelBase'?

File: ts_benchmark/models/advanced_model_base.py [hunk 3]
```diff
@@ -0,0 +1,724 @@
...
+        :return: The actual model object. The specific type of the return value should be defined by subclasses.
+        """
+        raise NotImplementedError("model must be implemented.")
+
+    def _adjust_lr(self, optimizer, epoch, config):
+        """
+        Adjusts the learning rate of the optimizer based on the current epoch and configuration.
+
+        This function is typically called during training to update the learning rate
+        according to a predefined schedule or strategy defined in the [adjust_learning_rate](file:///Users/pandajunkai/project/python/benchamrk/TFB/ts_benchmark/baselines/utils.py#L15-L32) function.
```
Comment (luckiezhou) [inline ts_benchmark/models/advanced_model_base.py:90]: use a relative path here

File: ts_benchmark/models/advanced_model_base.py [hunk 4]
```diff
@@ -0,0 +1,724 @@
...
+from torch.utils.data import DataLoader
+
+from ts_benchmark.baselines.utils import EarlyStopping, adjust_learning_rate
+from ts_benchmark.baselines.utils import (
+    forecasting_data_provider,
+    train_val_split,
+    get_time_mark,
+)
+from ts_benchmark.utils.data_processing import split_time
+from .model_base import ModelBase, BatchMaker
```
Comment (luckiezhou) [inline ts_benchmark/models/advanced_model_base.py:19]: I recommend to use absolute import in all cases

**Comment by luckiezhou** [comment]
some notes:
- please make sure there are enough tests before applying such a huge refactorization
- please recheck the documentation of the base class to see if they are clear enough, which is crutial for developers to implement new algorithms

future suggestions:
- you are putting your hands on the architecture design now, I recommend to refer to some commonly-used libraries, such as lightning, to refine your design whenever you are unsure about something
- as the repo becomes more complicated, it is better to build a standard CI/CD pipeline for code contribution (basically, a standard pipeline of tests). As it is resource consuming to test on deep models, at least you should have such a pipeline in code, and ensure the developers runs it locally before submitting.

File: ts_benchmark/baselines/utils.py [hunk 5]
```diff
@@ -12,6 +13,60 @@
 from ts_benchmark.utils.data_processing import split_time
 
 
+def adjust_learning_rate(optimizer, epoch, args):
+    # lr = args.learning_rate * (0.2 ** (epoch // 2))
+    if args.lradj == 'type1':
```
[the remark this answers is not in the document]
Comment (PandaJunk) [reply ts_benchmark/baselines/utils.py]: rename the four lr schedulers
截屏2025-06-21 23 22 22

File: ts_benchmark/baselines/utils.py (hunk 1 above)
Comment (PandaJunk) [reply ts_benchmark/baselines/utils.py:71]: add save_checkpoint function to base model which can be implemented by developers 
截屏2025-06-21 23 26 23

File: ts_benchmark/models/advanced_model_base.py (hunk 2 above)
Comment (PandaJunk) [reply ts_benchmark/models/advanced_model_base.py]: rename it to DeepForecastingModelBase which emphasizes the model is based on deep learning method in forecasting tasks
截屏2025-06-21 23 31 36

File: ts_benchmark/models/advanced_model_base.py (hunk 3 above)
Comment (PandaJunk) [reply ts_benchmark/models/advanced_model_base.py]: modify the above comment to:

This method is typically called to update the learning rate according to a predefined schedule.

File: ts_benchmark/models/advanced_model_base.py (hunk 4 above)
Comment (PandaJunk) [reply ts_benchmark/models/advanced_model_base.py]: now the import codes are similar with the following code:
截屏2025-06-21 23 40 14
